third_party/move/move-compiler-v2/src/pipeline/control_flow_graph_simplifier.rs

```diff
@@ -0,0 +1,687 @@
+//!
+//!     xx
+//!     yy
+//!
+//!     and block 2 is removed.
+//!
+//! 3. Removes all jumps to the next instruction:
+//!     Whenever we have
+//!
+//!     goto L;
+//!     L;
+//!
+//!     replace it by simply
+//!
+//!     L;
+//!
+//! 4. Replaces branch to same target by jump:
+//!     Whenever we have
+//!
+//!     if ... goto L else goto L
+//!
+//!     replace it by simply
+//!
+//!     goto L
+//!
+//! Side effects: removess all annotations.
+
+use itertools::Itertools;
+use move_model::model::FunctionEnv;
+use move_stackless_bytecode::{
+    function_target::FunctionData,
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    stackless_bytecode::{AbortAction, Bytecode, Label},
+    stackless_control_flow_graph::{BlockContent, BlockId, DFSLeft, StacklessControlFlowGraph},
+};
+use std::collections::{BTreeMap, BTreeSet};
+
+/// Simplifies the control flow graph as described in the module doc
+pub struct ControlFlowGraphSimplifier {}
+
+impl FunctionTargetProcessor for ControlFlowGraphSimplifier {
+    /// Does the first four transformations described in the module doc
+    fn process(
+        &self,
+        _targets: &mut FunctionTargetsHolder,
+        fun_env: &FunctionEnv,
+        data: FunctionData,
+        _scc_opt: Option<&[FunctionEnv]>,
+    ) -> FunctionData {
+        if fun_env.is_native() {
+            return data;
+        }
+        let mut transformer = ControlFlowGraphSimplifierTransformation::new(data);
+        let mut code_len = transformer.data.code.len();
+        // eliminates the tricky case where a block may contain repeated successor/predecessor blocks
+        transformer.eliminate_branch_to_same_target();
+        loop {
+            transformer.transform();
+            let new_code_len = transformer.data.code.len();
+            if new_code_len == code_len {
+                break;
+            } else {
+                code_len = new_code_len;
+            }
+        }
+        transformer.data
+    }
+
+    fn name(&self) -> String {
+        "ControlFlowGraphSimplifier".to_owned()
+    }
+}
+
+/// State of the transformation
+struct ControlFlowGraphSimplifierTransformation {
+    data: FunctionData,
+}
+
+impl ControlFlowGraphSimplifierTransformation {
+    pub fn new(data: FunctionData) -> Self {
+        Self { data }
+    }
+
+    /// Does the first four transformations described in the module doc
+    /// Requires: `eliminate_branch_to_same_target` has been called
+    fn transform(&mut self) {
+        let cfg1 = ControlFlowGraphCodeGenerator::new(std::mem::take(&mut self.data.code));
+        cfg1.check_consistency();
+        // may introduce new empty blocks, so perform before we remove empty blocks
+        let cfg2 = RedundantJumpRemover::transform(cfg1);
+        let cfg3 = EmptyBlockRemover::transform(cfg2);
+        if cfg!(debug_assertions) {
+            cfg3.check_consistency();
+        }
+        self.data.code = cfg3.gen_code(true);
```

Comment:
Why is `visit_all` true?

Reply:
If it is safe to do it here, let's do it. Unreachable code remover can be used for doing more in the future (like use const propagation results to eliminate more branches).

Reply:
done